### PR TITLE
kube-proxy: refactor UpdateServiceMap function as a method of ServiceMap

### PR DESCRIPTION
**Pull request by SataQiu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
kube-proxy: refactor UpdateServiceMap function as a method of ServiceMap
This ensures better encapsulation!
BTW, adjust the relative position of the `type` definition and its methods to improve the readability of our code.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

File: pkg/util/conntrack/conntrack.go [hunk 1]
```diff
@@ -114,7 +114,7 @@ func ClearEntriesForNAT(execer exec.Interface, origin, dest string, protocol v1.
 // https://github.com/kubernetes/kubernetes/issues/59368
 func ClearEntriesForPortNAT(execer exec.Interface, dest string, port int, protocol v1.Protocol) error {
 	if port <= 0 {
-		return fmt.Errorf("Wrong port number. The port number must be greater then zero")
+		return fmt.Errorf("Wrong port number. The port number must be greater than zero")
```
Comment (aojea) [inline pkg/util/conntrack/conntrack.go:117]: nit. this change seems unrelated, better have it split in a different commit addressing just the typo

**Comment by aojea** [comment]
/lgtm
it's also coherent with the `proxier.endpointsMap.Update` implementation

**Comment by SataQiu** [comment]
Kindly ping @thockin 
Does this look good to you?

File: pkg/proxy/service.go [hunk 2]
```diff
@@ -278,15 +278,15 @@ type UpdateServiceMapResult struct {
 	UDPStaleClusterIP sets.String
 }
 
-// UpdateServiceMap updates ServiceMap based on the given changes.
-func UpdateServiceMap(serviceMap ServiceMap, changes *ServiceChangeTracker) (result UpdateServiceMapResult) {
+// Update updates ServiceMap based on the given changes.
```
Comment (thockin) [inline pkg/proxy/service.go:281]: If we're going to do this, we should move the `type` definition above the method.

**Review by thockin — COMMENTED** [review]
Thanks for the cleanup.  That logic (change tracking) is pretty twisty and could really use better names and maybe better architecture :)

File: pkg/proxy/service.go (hunk 2 above)
Comment (SataQiu) [reply pkg/proxy/service.go]: Thanks for your guide @thockin . I found a similar problem in `pkg/proxy/endpoints.go`.
So I tried to adjust the relative position of the `type` definition and its methods.
IMO, this can improve the readability of our code. 
You are welcome to have another look :)

**Review by thockin — COMMENTED** [review]
Hi @SataQiu.  Thanks for this.  It SEEMS like a really nice PR, but I have to be honest - the diff makes it very very hard to review.  I don't mean to pick on you, but since you've been around a while, I hope you won't mind.  I'd like to use this PR as an example of what I call "prefactoring" - refactorings or cleanups that you can do to make a big diff smaller and easier to read.

I think you can take this main commit and break it into 4 or 5 MUCH smaller and easier to review commits.

If you need help with the git-fu to make this work, I am happy to help (inasmuch as I can :)

First: Add comments, fix typos, and do small cleanups in-place.  No large code changes, no moving blocks of code around.  These things should all be pretty trivial and easy to review.  "Oh, this just adds godoc comments", "this fixes an obvious mistake".  This can be one or many commits, but each diff should be trivially easy to verify.

Second: Rename things.  If you're going to rename functions or variables or whatever, do these in commits of their own.  This makes them trivial to review.  
                                                                                                                                             
Third: Move code around UNCHANGED.  If you want to move code around, either within a file or between files or between pkgs, do it here.  DO NOT CHANGE THE CODE OR TYPES.  Not even comments.  The only allowable change here is when you cross between pkgs or add tests.

Fourth: De-duplicate code.  If you are moving common code into a library, here is where you remove all the duplicates and fix the callers.

Fifth: Change function signatures. Convert functions to methods.  These changes are less trivial, but if you keep your commits small, the reviews are easy.

Sixth: Real work.  When I get to this point in your commit-stack, I already have a clue what you are up to.  These diffs should be free of collateral damage.  You already did the renames and comments and moving around before.  Each commit stands on its own, can be reviewed on its own, and "clears the path" for the final commit(s).

I think that this PR is a great candidate for prefactoring.  What do you think?

**Comment by SataQiu** [comment]
@thockin Thanks for your great guide! 
That is a good idea. And I'd love to have a try.

**Comment by thockin** [comment]
Thanks!  I have been looking for great examples of this, and I think we could turn this PR into that example.  Then I can link people to it and say "do this". :)
